app/src/behaviors/behavior_tri_state.c

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (c) 2022 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_tri_state
+
+#include <zephyr/device.h>
+#include <drivers/behavior.h>
+#include <zephyr/logging/log.h>
+#include <zmk/behavior.h>
+#include <zmk/behavior_queue.h>
+#include <zmk/keymap.h>
+#include <zmk/matrix.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/position_state_changed.h>
+#include <zmk/events/layer_state_changed.h>
+#include <zmk/hid.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#define ZMK_BHV_MAX_ACTIVE_TRI_STATES 10
+
+struct behavior_tri_state_config {
+    int32_t ignored_key_positions_len;
+    int32_t ignored_layers_len;
+    struct zmk_behavior_binding start_behavior;
+    struct zmk_behavior_binding continue_behavior;
+    struct zmk_behavior_binding end_behavior;
+    int32_t ignored_layers;
+    int32_t timeout_ms;
+    int tap_ms;
+    uint8_t ignored_key_positions[];
+};
+
+struct active_tri_state {
+    bool is_active;
+    bool is_pressed;
+    bool first_press;
+    uint32_t position;
+    const struct behavior_tri_state_config *config;
+    struct k_work_delayable release_timer;
+    int64_t release_at;
+    bool timer_started;
+    bool timer_cancelled;
+};
+
+static int stop_timer(struct active_tri_state *tri_state) {
+    int timer_cancel_result = k_work_cancel_delayable(&tri_state->release_timer);
+    if (timer_cancel_result == -EINPROGRESS) {
+        // too late to cancel, we'll let the timer handler clear up.
+        tri_state->timer_cancelled = true;
+    }
+    return timer_cancel_result;
+}
+
+static void reset_timer(int32_t timestamp, struct active_tri_state *tri_state) {
+    tri_state->release_at = timestamp + tri_state->config->timeout_ms;
+    int32_t ms_left = tri_state->release_at - k_uptime_get();
+    if (ms_left > 0) {
+        k_work_schedule(&tri_state->release_timer, K_MSEC(ms_left));
+        LOG_DBG("Successfully reset tri-state timer");
+    }
+}
+
+void trigger_end_behavior(struct active_tri_state *si) {
+    zmk_behavior_queue_add(si->position, si->config->end_behavior, true, si->config->tap_ms);
+    zmk_behavior_queue_add(si->position, si->config->end_behavior, false, 0);
+}
+
+void behavior_tri_state_timer_handler(struct k_work *item) {
+    struct active_tri_state *tri_state = CONTAINER_OF(item, struct active_tri_state, release_timer);
```

Comment:
from [Zephyr v3.5.0 Migration Guide](https://docs.zephyrproject.org/3.5.0/releases/migration-guide-3.5.html)

> [CONTAINER_OF](https://docs.zephyrproject.org/3.5.0/kernel/util/index.html#c.CONTAINER_OF) now performs type checking, this was very commonly misused to obtain user structure from [k_work](https://docs.zephyrproject.org/3.5.0/kernel/services/threads/workqueue.html#c.k_work) pointers without passing from [k_work_delayable](https://docs.zephyrproject.org/3.5.0/kernel/services/threads/workqueue.html#c.k_work_delayable). This would now result in a build error and have to be done correctly using [k_work_delayable_from_work()](https://docs.zephyrproject.org/3.5.0/kernel/services/threads/workqueue.html#c.k_work_delayable_from_work).

```suggestion
    struct k_work_delayable *ditem = k_work_delayable_from_work(item);
    struct active_tri_state *tri_state = CONTAINER_OF(ditem, struct active_tri_state, release_timer);
```